### PR TITLE
fixed unstaked list on delegation when nodes are unstaked from queue

### PR DESCRIPTION
**Pull request by sasurobert**
## Reasoning behind the pull request
Bugfix as nodes remain in undefined state from the delegation contract perspective after the activation of staking v4 and unStaking of nodes from queue

## Proposed changes
Called unStakeAtEndOfEpoch on delegation contracts to fix thee state of nodes.

## Testing procedure
Before activation of staking v4, create a delegation contract and add nodes to the queue. After the activation call reStakeUnStakedNodes and it should work

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/eei.go [hunk 1]
```diff
@@ -144,13 +144,17 @@ func (host *vmContext) GetStorageFromAddress(address []byte, key []byte) []byte
 		if value, isInMap := storageAdrMap[string(key)]; isInMap {
 			return value
 		}
+	} else {
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/eei.go:147]: I'm not sure I understand the reason behind these changes. They bring no utility from my point of view.
If you have **no storage** => you create empty map and fill it here : `storageAdrMap[string(key)] = data`, but this code has no effect on the internal data, nor the returned result. Why was this necessary?

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/multiversx/mx-chain-core-go/core"
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/stakingWaitingList.go:7]: Perhaps sort imports?

File: vm/systemSmartContracts/delegation.go [hunk 3]
```diff
@@ -2322,7 +2322,8 @@ func (d *delegation) deleteDelegatorIfNeeded(address []byte, delegator *Delegato
 }
 
 func (d *delegation) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	if !bytes.Equal(args.CallerAddr, d.endOfEpochAddr) {
+	if !bytes.Equal(args.CallerAddr, d.endOfEpochAddr) &&
+		!bytes.Equal(args.CallerAddr, d.stakingSCAddr) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/delegation.go:2326]: [question]
Also, is it really needed here to whitelist `stakingSCAddr`? 
In `unStakeAllNodesFromQueue` func, you call `unStakeAtEndOfEpoch` with caller address always `args.ReceipentAddr`, which is `vm.StakingSCAddress`. You could change there the caller address to be `EndOfEpochAddress`

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 4]
```diff
@@ -835,11 +838,42 @@ func (s *stakingSC) unStakeAllNodesFromQueue(args *vmcommon.ContractCallInput) v
...
 	}
 
 	// delete waiting list head element
 	s.eei.SetStorage([]byte(waitingListHeadKey), nil)
 
+	// call unStakeAtEndOfEpoch from the delegation contracts to compute the new unStaked list
+	for _, owner := range orderedListOwners {
+		listOfKeys := mapOwnerKeys[owner]
+
+		if s.eei.BlockChainHook().GetShardOfAddress([]byte(owner)) != core.MetachainShardId {
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/stakingWaitingList.go:859]: So we don't unstake nodes which have as owner an address? Why?

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 5]
```diff
@@ -835,11 +838,42 @@ func (s *stakingSC) unStakeAllNodesFromQueue(args *vmcommon.ContractCallInput) v
...
+			continue
+		}
+
+		unStakeCall := "unStakeAtEndOfEpoch"
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/stakingWaitingList.go:863]: We could extract this and the for below in a separate func to avoid `for in for`

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 6]
```diff
@@ -835,11 +838,42 @@ func (s *stakingSC) unStakeAllNodesFromQueue(args *vmcommon.ContractCallInput) v
...
+		}
+
+		mapOwnerKeys[ownerAddr] = append(mapOwnerKeys[ownerAddr], blsKey)
 	}
 
 	// delete waiting list head element
 	s.eei.SetStorage([]byte(waitingListHeadKey), nil)
 
+	// call unStakeAtEndOfEpoch from the delegation contracts to compute the new unStaked list
+	for _, owner := range orderedListOwners {
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/stakingWaitingList.go:856]: [question]
Why is it mandatory to unstake them ordered?

File: vm/systemSmartContracts/delegation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2326]: If I change the message it is backward incompatible.

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2326]: I do not want to fake the line of execution. it was called by the stakingSCAddr.

File: vm/systemSmartContracts/eei.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:147]: this is an optimization. I just saw that it could be added here - it can be reverted.

Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:147]: You need to create the internal map, otherwise it is nil. and yields in error.

File: vm/systemSmartContracts/stakingWaitingList.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:859]: this calls the delegation SCs. we need to update the list there.

Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:859]: UnStake of nodes have already happened a few lines before.

File: vm/systemSmartContracts/stakingWaitingList.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:856]: In order to have a deterministic execution.

File: vm/systemSmartContracts/stakingWaitingList.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:863]: done

File: vm/systemSmartContracts/stakingWaitingList.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:7]: done

File: vm/systemSmartContracts/delegation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:2326]: wondering if this is still backward compatible

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 7]
```diff
@@ -835,14 +837,55 @@ func (s *stakingSC) unStakeAllNodesFromQueue(args *vmcommon.ContractCallInput) v
...
+			orderedListOwners = append(orderedListOwners, ownerAddr)
+		}
+
+		mapOwnerKeys[ownerAddr] = append(mapOwnerKeys[ownerAddr], blsKey)
 	}
 
 	// delete waiting list head element
 	s.eei.SetStorage([]byte(waitingListHeadKey), nil)
 
+	// call unStakeAtEndOfEpoch from the delegation contracts to compute the new unStaked list
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/stakingWaitingList.go:854]: instead of the comment, we should split this into a new function. Related to @mariusmihaic comment.

Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:854]: already made a split inside.

File: integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	reStakeTxData := fmt.Sprintf("reStakeUnStakedNodes@%s", blsKeys[0])
+	reStakeNodes := generateTransaction(validatorOwner.Bytes, ownerNonce, delegationAddressBytes, big.NewInt(0), reStakeTxData, gasLimitForStakeOperation)
+	reStakeTx, err := cs.SendTxAndGenerateBlockTilTxIsExecuted(reStakeNodes, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, reStakeTx)
+
+	status = getBLSKeyStatus(t, metachainNode, decodedBLSKey0)
+	require.Equal(t, "staked", status)
+
+	err = cs.GenerateBlocks(20)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go:127]: from linter: ineffectual assignment to err (ineffassign)

File: integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	status := getBLSKeyStatus(t, metachainNode, decodedBLSKey0)
+	require.Equal(t, "queued", status)
+
+	// activate staking v4
+	err = cs.GenerateBlocksUntilEpochIsReached(int32(stakingV4ActivationEpoch))
+	require.Nil(t, err)
+
+	status = getBLSKeyStatus(t, metachainNode, decodedBLSKey0)
+	require.Equal(t, "unStaked", status)
+
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go:116]: Please add here a VM query to the delegation contract, `getAllNodeStates` function

File: integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go (hunk 8 above)
Comment (miiu96) [reply integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go:127]: fixed

File: integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
+package staking
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/multiversx/mx-chain-core-go/core"
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/chainSimulator/staking/stakingProviderWithNodesinQueue_test.go:10]: fixeds

**Review by iulianpascalau — APPROVED** [review]

File: vm/systemSmartContracts/staking_test.go [hunk 11]
```diff
@@ -3706,3 +3706,107 @@ func TestStakingSc_UnStakeAllFromQueue(t *testing.T) {
...
+	enableEpochsHandler.AddActiveFlags(common.StakeFlag)
+
+	// do stake should work
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("firstKey "))
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("secondKey"))
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("thirdKey "))
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("fourthKey"))
+
+	waitingReturn := doGetWaitingListRegisterNonceAndRewardAddress(t, stakingSmartContract, eei)
+	assert.Equal(t, len(waitingReturn), 9)
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/staking_test.go:3758]: Why 9? You only staked 4 keys

File: vm/systemSmartContracts/staking_test.go [hunk 12]
```diff
@@ -3706,3 +3706,107 @@ func TestStakingSc_UnStakeAllFromQueue(t *testing.T) {
...
+	assert.Equal(t, retCode, vmcommon.Ok)
+
+	assert.Equal(t, len(eei.GetStorage([]byte(waitingListHeadKey))), 0)
+	newHead, _ := stakingSmartContract.getWaitingListHead()
+	assert.Equal(t, uint32(0), newHead.Length) // no entries in the queue list
+
+	marshaledData = eei.GetStorageFromAddress(stakerAddress, []byte(delegationStatusKey))
+	_ = stakingSmartContract.marshalizer.Unmarshal(dStatus, marshaledData)
+	assert.Equal(t, len(dStatus.UnStakedKeys), 3)
+	assert.Equal(t, len(dStatus.StakedKeys), 1)
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/staking_test.go:3801]: Why 3 unstaked? you had 4 keys in waiting list

File: vm/systemSmartContracts/staking_test.go [hunk 13]
```diff
@@ -3706,3 +3706,107 @@ func TestStakingSc_UnStakeAllFromQueue(t *testing.T) {
...
+
+	doGetStatus(t, stakingSmartContract, eei, []byte("secondKey"), "unStaked")
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/staking_test.go:3803]: We could check all keys status here

File: vm/systemSmartContracts/staking_test.go [hunk 14]
```diff
@@ -3706,3 +3706,107 @@ func TestStakingSc_UnStakeAllFromQueue(t *testing.T) {
...
+
+	// stake them again - as they were deleted from waiting list
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("thirdKey "))
+	doStake(t, stakingSmartContract, stakingAccessAddress, stakerAddress, []byte("fourthKey"))
+
+	// surprisingly, the queue works again as we did not activate the staking v4
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/staking_test.go:3809]: ?
You activated staking v4 and there is no queue

File: vm/systemSmartContracts/staking_test.go (hunk 11 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go:3758]: changed the assert

File: vm/systemSmartContracts/staking_test.go (hunk 13 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go:3803]: added

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
Normal allin test: v1.6.15-dev-config-eb2e06c06d -> correct-unStakedList-after-99c9793df2

--- Specific errors ---

block hash does not match 5003
wrong nonce in block 2031
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 1
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  1
Nr. of all WARNS:  529
Nr. of new ERRORS:  1
Nr. of new WARNS:  10
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/
